### ISSUE TITLE
[Suggestion] Add the ability to control the generation of ores.

### ISSUE THREAD

**Issue by Tsyklop** [ReikaKalseki/Reika_Mods_Issues#240]
Please add the config line which are responsible for the generation of a particular ore in the world, that we could adjust the number as well as turn off all ...

**Comment by ReikaKalseki** [comment]
No. This has been discussed ad nauseam. You have options to discretize it, or disable it if another mod is present. I am not allowing total removal of the ore with no replacement.

**Comment by Tsyklop** [comment]
@ReikaKalseki Where can I find this option?("You have options to discretize it")

**Comment by ReikaKalseki** [comment]
Where all the other options are. Also, are you sure you know what discretize means in this context?

**Comment by Tsyklop** [comment]
@ReikaKalseki Not sure. Tell me please.

**Comment by ReikaKalseki** [comment]
It generates ore in N times fewer chunks, but at N times the density within those chunks, essentially (crudely) imitating a more "COG/GT/Ore Vein" style generation.

**Comment by Tsyklop** [comment]
@ReikaKalseki And where you can change this value?("You have options to discretize it")

**Comment by ljfa-ag** [comment]
In the config folder in the Reika subfolder. There you can change it in ElectriCraft.cfg and ReactorCraft.cfg.

**Comment by Tsyklop** [comment]
@ljfa-ag Thx. 
I:"Ore Discretization"=1
this? to disable the need to put a "0"?

**Comment by ljfa-ag** [comment]
No, it can't be disabled completely if you have no other mod generating it. Read what Reika wrote above.

**Comment by Tsyklop** [comment]
@ljfa-ag I stand mods to replace all ores.

**Comment by Tsyklop** [comment]
I installed "I:" Ore Discretization "= 0" and received a crash during the generation of the world.

What is the need to put it?

**Comment by ReikaKalseki** [comment]
You were told that zero would not work and did it anyways. Why are you asking for help with that?

**Comment by Tsyklop** [comment]
because I interfere with ore from the Reactor Craft and I'm looking for a way to remove them from generation ...

**Comment by ReikaKalseki** [comment]
And you have been told something like six times that that was not possible. Why does your failure to succeed at it puzzle you?

**Comment by Tsyklop** [comment]
I understood. Close this question ..

**Comment by ReikaKalseki** [comment]
I closed it almost a week ago.

**Comment by ReikaKalseki** [comment]
Also, for what it is worth, as of v9 ReC will crash on load if you specify zero or > 100 for the discretization option. Zero leads to a divide-by-zero error, and > 100 will lag intensely every few dozen chunks.

**Comment by TomeWyrm** [comment]
@Tsyklop use Custom Ore Generation

**Comment by Tsyklop** [comment]
@TomeWyrm I'll try.

**Comment by Haggle1996** [comment]
Can I have my lag back?

I'm not fond of this change. Previously, I used Int.Max to simulate disabling the ores (and replacing them with CoFHCore's custom world gen in order to get rich, rare veins). With ore descretization now capped at 100, I end up with a huge, ugly column of ore every 10x10 chunks, and leaving it at default ruins the feel of the custom ore gen.

EDIT: v9 doesn't affect world gen performance

So, please reconsider? I'm really not fond of the idea of rewriting my entire custom ore generation using a different mod, particularly when I have so many users playing the existing pack.

**Comment by TomeWyrm** [comment]
If you're using CoG, just use it to remove Reika's ores.

**Comment by Tsyklop** [comment]
@TomeWyrm it do not understand how to remove the ore.

**Comment by ReikaKalseki** [comment]
> CI'm not fond of this change. Previously, I used Int.Max to simulate disabling the ores (and replacing them with CoFHCore's custom world gen in order to get rich, rare veins). With ore descretization now capped at 100, I end up with a huge, ugly column of ore every 10x10 chunks, and leaving it at default ruins the feel of the custom ore gen.

At Int.MAX, you are literally asking for random crashes. And whatever you do or say regarding knowing the risks and accepting them, it is irrelevant: For every person like you there are fifty more who will _happily_ pick absurd settings and then complain.

**Comment by TomeWyrm** [comment]
I'm the wrong person to walk you through, and this is the wrong place.

http://www.minecraftforum.net/forums/mapping-and-modding/minecraft-mods/1291128-custom-ore-generation-first-revival

If you're having issues, ask there. There is actually good documentation, but the thread is friendly.

**Comment by Haggle1996** [comment]
@ReikaKalseki then please, give me another option to make this work out. I try excessively hard to balance Revolution around your mods. Your mods ARE the reason for my pack. Time and again, I field complaints on your behalf. For example, I get complaints about HSLA armor being to easy to craft and that I've disabled vanilla armors in my pack. I disabled them precisely because HSLA is easy to get, and I _want_ people to work their way through RotaryCraft. Mobs in the pack are balanced around the fact that HSLA is easy to get (and rewards people for getting to bedrock armor).

With the oregen in my pack today (pre-v9), finding an ore while spelunking is a happy event. 'ooh! What did I find?!?!?' It's a pleasing event. When the world is scattered full of fluorite, it makes discovery feel very...disappointing. It's going to increase unhappiness with oregen, which is a delight for virtually all players today (large, rare veins -- takes a while to find something, but when you do, you find plenty to last for a while -- somewhat like GT, but without as much grind).

I try very hard to maintain your vision for your mods -- I have less than zero desire to interfere with your artistic vision, and if anything I want to promote it more, so that others experience your genius. Depending on the day, hundreds, if not thousands of players interact with RotaryCraft through Revolution. On a typical day, 500 to 700 NEW players install my pack. I have no idea how many are new to your mods, but please, don't make it a pain in my arse to keep things maintained going forward.

**Comment by ReikaKalseki** [comment]
This has nothing to do with artistic vision. This is about protecting me from bug reports.

One thing you _could_ do is have a dedicated ore-removal mod...

**Comment by Haggle1996** [comment]
Uh huh. Write a mod that removes the ore you generate after you generate it but before a custom ore gen mod re-adds it back? Wouldn't it be easier for you to add a functional config option like below? I'd be happy to write the pull request. It would satisfy a very long and outstanding request from your user base.

# Public packs that turn off ore generation must add an alternate method of ore generation to retain distribution permissions
B:DisableOreGeneration=true

**Comment by Tsyklop** [comment]
I will add to what has been said above.

Do not understand how to clean generation using Custom OreGen.
It would be much wiser to make the function to disable the generation of ...
It is flexible for developers mod packs, as well as it saves you from reading these same proposals and discussions in the future.

**Comment by ReikaKalseki** [comment]
"After the fact" enforcement like it being required for pack permissions is both ineffective at protecting me - most of the issues come earlier than that - as well as only compounding the "oh look at Reika being so controlling" problems I have.

I need something that only someone capable can modify.

**Comment by TomeWyrm** [comment]
Post in the CoG thread. What you BOTH want to do is VERY easy to do with CoG; no joke.

**Comment by Haggle1996** [comment]
I'm not disputing that CoG might address the issue. My issue is that I have spent weeks, if not months, of effort fine-tuning and play testing the ore gen in my pack using CoFH. Getting it 'just right' in CoG is effort I'm not looking forward to, particularly when it was working previously.

**Comment by ReikaKalseki** [comment]
Do you have any ideas for my system?

**Comment by Haggle1996** [comment]
Nothing constructive yet. I do see where you're coming from, so trying to brainstorm a solution before I post another non-starter. :)

**Comment by BarbasTheDog** [comment]
If this is still up in the air, may I suggest a solution?

How about making something like _B:DisableOreGeneration=true_ valid, but make the line not generate with the default configs? That way, if someone just up and decides to mess around with them, they won't be able to disable ore gen. If they care enough to ask you for a way to do so, though, they probably understand what that means and what the consequences are, so you can just make sure they understand and point out that adding the line is a valid option.

**Comment by ReikaKalseki** [comment]
> How about making something like B:DisableOreGeneration=true valid, but make the line not generate with the default configs?

That is not how the Forge config system works. Any option provided automatically generates.

**Comment by Haggle1996** [comment]
On a whim, this weekend, I rewrote a good portion of my packs world gen using Custom Ore Gen mentioned above. Ore gen in my pack went from 'painfully slow' to 'I would rather pass kidney stones for a living'...almost three times slower than CoFH. But the veins look pretty.

@BarbasTheDog MCF doesn't allow for hidden config options. If you specify the configuration option, the default value is written to a config file. That said, a similar approach with a Json file?

**Comment by BarbasTheDog** [comment]
> @BarbasTheDog MCF doesn't allow for hidden config options. If you specify the configuration option, the default value is written to a config file. That said, a similar approach with a Json file?

Or even just checking during init if the line was added. Sorta hacky? Sure, but it'd work.

**Comment by Haggle1996** [comment]
Reika was trying to point out that missing options are automatically added to Forge config files. So checking for a missing line would be pointless -- Forge will add the config line if its missing.

**Comment by BarbasTheDog** [comment]
I meant checking the file directly in the same class and just before the ores for oregen are registered - I understand that forge's normal config system would do that, but the mod can access the file system, right? (I'm sorry, I'm not very familiar with Java)

Like I said, kind of dirty, but probably effective.

**Comment by aurumvorax** [comment]
This is a change I desperately want for my modpack as well, so I've got a suggestion.  It's a bit hacky, but not as bad as having a seperate json.  Add say:
B:DisableOreGen = true
I:OreGenPassword = 000000

and hide the actual password in the source code?

Would that offer the protection you are looking for Reika?

**Comment by ReikaKalseki** [comment]
No. For one, all it takes is someone to start spreading it and it no longer serves its intended purpose, and I cannot "unbreak" it without simultaneously breaking every pack. Second, it is not required to understand Java to be able to dig the password off of the GitHub source.

**Comment by Haggle1996** [comment]
@aurumvorax use a mod called 'Custom Ore Gen Revival', suggested earlier on this thread. Out of the box, it has many options for customizing ore generation. It's easy to remove any ores you don't want.

**Comment by TomeWyrm** [comment]
The problem is a pretty damn large performance hit during worldgen

**Comment by aurumvorax** [comment]
@Reika Fair enough, I suppose I was thinking of something more than a hardcoded password, but yeah, I can see the problem with people spreading it around.  What level of capability would you consider sufficient to allow disabling oregen?  Alternatively, would it be feasible to check if another mod is genning ReC/ElC ores instead?

@TomeWyrm I have looked at COG, and while it can sortof do what I want, it adds an order of magnitute to worldgen times, which are already too long with my pack.

**Comment by cosmicdan** [comment]
Reika, how about disabling oregen if the host system has a specific environment variable set? You could just use `if (System.getenv("REIKA_OREGEN").equals("DISABLED")); return` somewhere in your code. I can't think of how this helps redistribution of existing modpacks ( sorry @Haggle1996 ) but this does a good job of ensuring protection from false bug reports for you, yet still allows servers to disable it.

Modpack makers would have to ask all their users to manually set this on their host OS. As far as I know, all the modpack launchers out there don't allow any such configuration. MultiMC could do it via Custom Commands, I think, but even if that was distributed with modpacks the custom command would vary between Windows/Unix.

TL;DR: Use an environment variable. As far as I know this would require the users to manually disable oregen on their PC (or server).

Please? :3

**Comment by ReikaKalseki** [comment]
I have no intention of giving servers the ability to randomly disable it, especially given the much higher chance of it being done so stupidly if done by servers as opposed to pack developers. Also, the moment I start checking things like that, I start having to deal with people like this:

https://www.reddit.com/r/feedthebeast/comments/4d317f/v12ev13_next_update_notice_config_system/d1nptsh

https://www.reddit.com/r/feedthebeast/comments/3m0bqk/this_seems_like_a_terrible_terrible_idea_remote/

http://forum.feed-the-beast.com/threads/recent-events-discussion-red-thread.45220/page-209#post-1116824

https://www.reddit.com/r/feedthebeast/comments/4o0bkw/ies_failing_to_launch_because_a_mod_developers/d490t5h

**Comment by cosmicdan** [comment]
Alright, that's reasonable logic. COG it is then.

On 16 Jun 2016, Reika notifications@github.com wrote:

> I have no intention of giving servers the ability to randomly disable
> it, especially given the much higher chance of it being done so
> stupidly if done by servers as opposed to pack developers. Also, the
> moment I start checking things like that, I start having to deal with
> people like this:
> 
> https://www.reddit.com/r/feedthebeast/comments/4d317f/v12ev13_next_update_notice_config_system/d1nptsh
> 
> https://www.reddit.com/r/feedthebeast/comments/3m0bqk/this_seems_like_a_terrible_terrible_idea_remote/
> 
> http://forum.feed-the-beast.com/threads/recent-events-discussion-red-thread.45220/page-209#post-1116824
> 
> https://www.reddit.com/r/feedthebeast/comments/4o0bkw/ies_failing_to_launch_because_a_mod_developers/d490t5h
> 
> ---
> 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/ReikaKalseki/Reika_Mods_Issues/issues/240#issuecomment-226203926

## 

Sent with <a href="https://play.google.com/store/apps/details?id=com.onegravity.k10.pro2"><b>K-@ Mail</b></a> - the evolution of emailing.
